File: sdks/sdk-go/sigilum/http_signatures.go
package sigilum

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	neturl "net/url"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var signatureInputPattern = regexp.MustCompile(`^sig1=\(([^)]*)\);created=(\d+);keyid="([^"]+)";alg="([^"]+)";nonce="([^"]+)"$`)
var signaturePattern = regexp.MustCompile(`^sig1=:([^:]+):$`)

var requiredComponentSetsNoBody = [][]string{
	{
		"@method",
		"@target-uri",
		"sigilum-namespace",
		"sigilum-subject",
		"sigilum-agent-key",
		"sigilum-agent-cert",
	},
	{
		"@method",
		"@target-uri",
		"sigilum-namespace",
		"sigilum-subject",
		"sigilum-agent-id",
		"sigilum-agent-key",
		"sigilum-agent-cert",
	},
}

var requiredComponentSetsWithBody = [][]string{
	{
		"@method",
		"@target-uri",
		"content-digest",
		"sigilum-namespace",
		"sigilum-subject",
		"sigilum-agent-key",
		"sigilum-agent-cert",
	},
	{
		"@method",
		"@target-uri",
		"content-digest",
		"sigilum-namespace",
		"sigilum-subject",
		"sigilum-agent-id",
		"sigilum-agent-key",
		"sigilum-agent-cert",
	},
}

const (
	verifyCodeMissingSignatureHeaders   = "SIG_MISSING_SIGNATURE_HEADERS"
	verifyCodeSignatureInputInvalid     = "SIG_SIGNATURE_INPUT_INVALID"
	verifyCodeAlgorithmUnsupported      = "SIG_ALGORITHM_UNSUPPORTED"
	verifyCodeCreatedInvalid            = "SIG_CREATED_INVALID"
	verifyCodeTimestampOutOfRange       = "SIG_TIMESTAMP_OUT_OF_RANGE"
	verifyCodeReplayDetected            = "SIG_REPLAY_DETECTED"
	verifyCodeSignatureHeaderInvalid    = "SIG_SIGNATURE_HEADER_INVALID"
	verifyCodeCertHeaderMissing         = "SIG_CERT_HEADER_MISSING"
	verifyCodeCertHeaderInvalid         = "SIG_CERT_HEADER_INVALID"
	verifyCodeCertInvalid               = "SIG_CERT_INVALID"
	verifyCodeNamespaceMismatch         = "SIG_NAMESPACE_MISMATCH"
	verifyCodeExpectedNamespaceMismatch = "SIG_EXPECTED_NAMESPACE_MISMATCH"
	verifyCodeSubjectMissing            = "SIG_SUBJECT_MISSING"
	verifyCodeExpectedSubjectMismatch   = "SIG_EXPECTED_SUBJECT_MISMATCH"
	verifyCodeSubjectComponentMissing   = "SIG_SUBJECT_COMPONENT_MISSING"
	verifyCodeSignedComponentsInvalid   = "SIG_SIGNED_COMPONENTS_INVALID"
	verifyCodeKeyHeaderMissing          = "SIG_KEY_HEADER_MISSING"
	verifyCodeKeyMismatch               = "SIG_KEY_MISMATCH"
	verifyCodeKeyIDMismatch             = "SIG_KEY_ID_MISMATCH"
	verifyCodeContentDigestMismatch     = "SIG_CONTENT_DIGEST_MISMATCH"
	verifyCodeSigningBaseInvalid        = "SIG_SIGNING_BASE_INVALID"
	verifyCodeKeyHeaderInvalid          = "SIG_KEY_HEADER_INVALID"
	verifyCodeVerificationFailed        = "SIG_VERIFICATION_FAILED"
)

type parsedSignatureInput struct {
	Components []string
	Created    int64
	KeyID      string
	Alg        string
	Nonce      string
}

func normalizeHeaders(headers map[string]string) map[string]string {
	out := map[string]string{}
	for key, value := range headers {
		out[strings.ToLower(strings.TrimSpace(key))] = strings.TrimSpace(value)
	}
	return out
}

func normalizeMethod(method string) string {
	if strings.TrimSpace(method) == "" {
		return "GET"
	}
	return strings.ToUpper(strings.TrimSpace(method))
}

func contentDigest(body []byte) string {
	sum := sha256.Sum256(body)
	return "sha-256=:" + base64.StdEncoding.EncodeToString(sum[:]) + ":"
}

func signatureParams(components []string, created int64, keyID, nonce string) string {
	quoted := make([]string, 0, len(components))
	for _, component := range components {
		quoted = append(quoted, "\""+component+"\"")
	}
	return "(" + strings.Join(quoted, " ") + ");created=" +
		strconv.FormatInt(created, 10) +
		";keyid=\"" + keyID + "\";alg=\"ed25519\";nonce=\"" + nonce + "\""
}

func normalizeTargetURI(raw string) string {
	parsed, err := neturl.Parse(raw)
	if err != nil {
		return raw
	}
	parsed.Fragment = ""
	return parsed.String()
}

func componentValue(component string, method string, url string, headers map[string]string) (string, error) {
	switch component {
	case "@method":
		return strings.ToLower(method), nil
	case "@target-uri":
		return normalizeTargetURI(url), nil
	default:
		value, ok := headers[component]
		if !ok || strings.TrimSpace(value) == "" {
			return "", fmt.Errorf("missing required signed header: %s", component)
		}
		return value, nil
	}
}

func signingBase(components []string, method string, url string, headers map[string]string, sigParams string) ([]byte, error) {
	lines := make([]string, 0, len(components)+1)
	for _, component := range components {
		value, err := componentValue(component, method, url, headers)
		if err != nil {
			return nil, err
		}
		lines = append(lines, fmt.Sprintf("\"%s\": %s", component, value))
	}
	lines = append(lines, fmt.Sprintf("\"@signature-params\": %s", sigParams))
	return []byte(strings.Join(lines, "\n")), nil
}

func EncodeCertificateHeader(certificate SigilumCertificate) (string, error) {
	bytes, err := json.Marshal(certificate)
	if err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(bytes), nil
}

func DecodeCertificateHeader(value string) (SigilumCertificate, error) {
	bytes, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		return SigilumCertificate{}, err
	}
	var certificate SigilumCertificate
	if err := json.Unmarshal(bytes, &certificate); err != nil {
		return SigilumCertificate{}, err
	}
	return certificate, nil
}

func parseSignatureInputHeader(value string) (parsedSignatureInput, error) {
	match := signatureInputPattern.FindStringSubmatch(strings.TrimSpace(value))
	if len(match) != 6 {
		return parsedSignatureInput{}, fmt.Errorf("invalid Signature-Input header format")
	}
	componentsRaw := match[1]
	createdRaw := match[2]
	keyID := match[3]
	alg := match[4]
	nonce := match[5]

	created, err := strconv.ParseInt(createdRaw, 10, 64)
	if err != nil {
		return parsedSignatureInput{}, fmt.Errorf("invalid created timestamp in Signature-Input")
	}

	parts := strings.Fields(strings.TrimSpace(componentsRaw))
	components := make([]string, 0, len(parts))
	for _, part := range parts {
		if len(part) < 2 || !strings.HasPrefix(part, "\"") || !strings.HasSuffix(part, "\"") {
			return parsedSignatureInput{}, fmt.Errorf("invalid component in Signature-Input: %s", part)
		}
		components = append(components, strings.TrimSuffix(strings.TrimPrefix(part, "\""), "\""))
	}

	return parsedSignatureInput{
		Components: components,
		Created:    created,
		KeyID:      keyID,
		Alg:        alg,
		Nonce:      nonce,
	}, nil
}

func parseSignatureHeader(value string) ([]byte, error) {
	match := signaturePattern.FindStringSubmatch(strings.TrimSpace(value))
	if len(match) != 2 {
		return nil, fmt.Errorf("invalid Signature header format")
	}
	return base64.StdEncoding.DecodeString(match[1])
}

func SignHTTPRequest(identity SigilumIdentity, input SignRequestInput) (SignedRequest, error) {
	if strings.TrimSpace(input.URL) == "" {
		return SignedRequest{}, fmt.Errorf("request URL is required")
	}
	method := normalizeMethod(input.Method)
	normalizedURL := normalizeTargetURI(input.URL)
	headers := normalizeHeaders(input.Headers)
	body := input.Body

	if len(body) > 0 {
		headers["content-digest"] = contentDigest(body)
	}

	certHeader, err := EncodeCertificateHeader(identity.Certificate)
	if err != nil {
		return SignedRequest{}, err
	}

	headers["sigilum-namespace"] = identity.Namespace
	subject := strings.TrimSpace(input.Subject)
	if subject == "" {
		if existing := strings.TrimSpace(headers["sigilum-subject"]); existing != "" {
			subject = existing
		} else {
			subject = identity.Namespace
		}
	}
	headers["sigilum-subject"] = subject
	headers["sigilum-agent-key"] = identity.PublicKey
	headers["sigilum-agent-cert"] = certHeader

	components := []string{"@method", "@target-uri", "sigilum-namespace", "sigilum-subject", "sigilum-agent-key", "sigilum-agent-cert"}
	if len(body) > 0 {
		components = []string{"@method", "@target-uri", "content-digest", "sigilum-namespace", "sigilum-subject", "sigilum-agent-key", "sigilum-agent-cert"}
	}

	created := input.Created
	if created <= 0 {
		created = time.Now().Unix()
	}
	nonce := strings.TrimSpace(input.Nonce)
	if nonce == "" {
		nonce, err = randomUUIDV4()
		if err != nil {
			return SignedRequest{}, err
		}
	}
	sigParams := signatureParams(components, created, identity.KeyID, nonce)
	base, err := signingBase(components, method, normalizedURL, headers, sigParams)
	if err != nil {
		return SignedRequest{}, err
	}

	private := ed25519.NewKeyFromSeed(identity.PrivateKey)
	signature := ed25519.Sign(private, base)

	headers["signature-input"] = "sig1=" + sigParams
	headers["signature"] = "sig1=:" + base64.StdEncoding.EncodeToString(signature) + ":"

	return SignedRequest{
		URL:     normalizedURL,
		Method:  method,
		Headers: headers,
		Body:    body,
	}, nil
}

func VerifyHTTPSignature(input VerifySignatureInput) VerifySignatureResult {
	invalid := func(code string, reason string) VerifySignatureResult {
		return VerifySignatureResult{Valid: false, Code: code, Reason: reason}
	}

	headers := normalizeHeaders(input.Headers)
	signatureInput := headers["signature-input"]
	signatureValue := headers["signature"]

	if signatureInput == "" || signatureValue == "" {
		return invalid(verifyCodeMissingSignatureHeaders, "Missing Signature-Input or Signature header")
	}

	parsed, err := parseSignatureInputHeader(signatureInput)
	if err != nil {
		return invalid(verifyCodeSignatureInputInvalid, err.Error())
	}
	if strings.ToLower(parsed.Alg) != "ed25519" {
		return invalid(verifyCodeAlgorithmUnsupported, "Unsupported signature algorithm")
	}
	if parsed.Created <= 0 {
		return invalid(verifyCodeCreatedInvalid, "Invalid Signature-Input created timestamp")
	}
	if input.MaxAgeSeconds > 0 {
		now := input.NowUnix
		if now <= 0 {
			now = time.Now().Unix()
		}
		age := now - parsed.Created
		if age < 0 || age > input.MaxAgeSeconds {
			return invalid(verifyCodeTimestampOutOfRange, "Signature expired or not yet valid")
		}
	}
	if input.SeenNonces != nil {
		if _, exists := input.SeenNonces[parsed.Nonce]; exists {
			return invalid(verifyCodeReplayDetected, "Replay detected: nonce already seen")
		}
		input.SeenNonces[parsed.Nonce] = struct{}{}
	}

	signature, err := parseSignatureHeader(signatureValue)
	if err != nil {
		return invalid(verifyCodeSignatureHeaderInvalid, err.Error())
	}

	certHeader := headers["sigilum-agent-cert"]
	if certHeader == "" {
		return invalid(verifyCodeCertHeaderMissing, "Missing sigilum-agent-cert header")
	}
	certificate, err := DecodeCertificateHeader(certHeader)
	if err != nil {
		return invalid(verifyCodeCertHeaderInvalid, "Invalid sigilum-agent-cert header")
	}
	if !VerifyCertificate(certificate) {
		return invalid(verifyCodeCertInvalid, "Invalid agent certificate")
	}

	namespaceHeader := headers["sigilum-namespace"]
	if namespaceHeader == "" || namespaceHeader != certificate.Namespace {
		return invalid(verifyCodeNamespaceMismatch, "Namespace header mismatch")
	}
	if strings.TrimSpace(input.ExpectedNamespace) != "" && input.ExpectedNamespace != namespaceHeader {
		return invalid(
			verifyCodeExpectedNamespaceMismatch,
			fmt.Sprintf("Namespace mismatch: expected %s, got %s", input.ExpectedNamespace, namespaceHeader),
		)
	}
	subjectHeader := strings.TrimSpace(headers["sigilum-subject"])
	if subjectHeader == "" {
		return invalid(verifyCodeSubjectMissing, "Missing sigilum-subject header")
	}
	if strings.TrimSpace(input.ExpectedSubject) != "" && input.ExpectedSubject != subjectHeader {
		return invalid(
			verifyCodeExpectedSubjectMismatch,
			fmt.Sprintf("Subject mismatch: expected %s, got %s", input.ExpectedSubject, subjectHeader),
		)
	}
	if !containsComponent(parsed.Components, "sigilum-subject") {
		return invalid(verifyCodeSubjectComponentMissing, "Missing sigilum-subject in signed components")
	}
	hasBody := len(input.Body) > 0
	if !hasValidSignedComponentSet(parsed.Components, hasBody) {
		return invalid(verifyCodeSignedComponentsInvalid, "Invalid signed component set")
	}

	keyHeader := headers["sigilum-agent-key"]
	if keyHeader == "" {
		return invalid(verifyCodeKeyHeaderMissing, "Missing sigilum-agent-key header")
	}
	if keyHeader != certificate.PublicKey {
		return invalid(verifyCodeKeyMismatch, "Certificate public key mismatch")
	}
	if parsed.KeyID != certificate.KeyID {
		return invalid(verifyCodeKeyIDMismatch, "keyid mismatch")
	}

	if len(input.Body) > 0 {
		expectedDigest := contentDigest(input.Body)
		if headers["content-digest"] != expectedDigest {
			return invalid(verifyCodeContentDigestMismatch, "Content digest mismatch")
		}
	}

	sigParams := signatureParams(parsed.Components, parsed.Created, parsed.KeyID, parsed.Nonce)
	base, err := signingBase(parsed.Components, normalizeMethod(input.Method), normalizeTargetURI(input.URL), headers, sigParams)
	if err != nil {
		return invalid(verifyCodeSigningBaseInvalid, err.Error())
	}

	publicKey, err := publicKeyFromEncoded(keyHeader)
	if err != nil {
		return invalid(verifyCodeKeyHeaderInvalid, "Invalid sigilum-agent-key header")
	}

	if !ed25519.Verify(ed25519.PublicKey(publicKey), base, signature) {
		return invalid(verifyCodeVerificationFailed, "Signature verification failed")
	}

	return VerifySignatureResult{Valid: true, Namespace: certificate.Namespace, Subject: subjectHeader, KeyID: certificate.KeyID}
}

func containsComponent(components []string, value string) bool {
	target := strings.TrimSpace(value)
	if target == "" {
		return false
	}
	for _, component := range components {
		if strings.TrimSpace(component) == target {
			return true
		}
	}
	return false
}

func hasValidSignedComponentSet(components []string, hasBody bool) bool {
	expectedSets := requiredComponentSetsNoBody
	if hasBody {
		expectedSets = requiredComponentSetsWithBody
	}
	for _, expected := range expectedSets {
		if len(components) != len(expected) {
			continue
		}
		match := true
		for idx := range expected {
			if components[idx] != expected[idx] {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

func randomUUIDV4() (string, error) {
	buffer := make([]byte, 16)
	if _, err := rand.Read(buffer); err != nil {
		return "", err
	}
	buffer[6] = (buffer[6] & 0x0f) | 0x40
	buffer[8] = (buffer[8] & 0x3f) | 0x80
	hexValue := hex.EncodeToString(buffer)
	return hexValue[0:8] + "-" + hexValue[8:12] + "-" + hexValue[12:16] + "-" + hexValue[16:20] + "-" + hexValue[20:32], nil
}

File: sdks/sdk-ts/src/http-signatures.ts
import "./setup.js";
import * as ed from "@noble/ed25519";
import { createHash, randomUUID } from "node:crypto";
import { verifyCertificate } from "./identity-store.js";
import type {
  SigilumCertificate,
  SigilumIdentity,
  SignRequestInput,
  SignedRequest,
  VerifySignatureInput,
  VerifySignatureResult,
} from "./types.js";

type SignatureInputParts = {
  components: string[];
  created: number;
  keyId: string;
  alg: string;
  nonce: string;
};

const REQUIRED_COMPONENTS_NO_BODY = [
  "@method",
  "@target-uri",
  "sigilum-namespace",
  "sigilum-subject",
  "sigilum-agent-key",
  "sigilum-agent-cert",
] as const;

const REQUIRED_COMPONENTS_WITH_BODY = [
  "@method",
  "@target-uri",
  "content-digest",
  "sigilum-namespace",
  "sigilum-subject",
  "sigilum-agent-key",
  "sigilum-agent-cert",
] as const;

function normalizeUrl(input: string | URL): URL {
  const url = input instanceof URL ? new URL(input.toString()) : new URL(input);
  url.hash = "";
  return url;
}

function normalizeBody(
  body: SignRequestInput["body"],
): Uint8Array | null {
  if (body == null) {
    return null;
  }
  if (typeof body === "string") {
    return new TextEncoder().encode(body);
  }
  if (body instanceof Uint8Array) {
    return body;
  }
  if (body instanceof ArrayBuffer) {
    return new Uint8Array(body);
  }

  throw new Error(
    "Unsupported body type for signing. Use string, Uint8Array, ArrayBuffer, or null.",
  );
}

function buildSignatureParams(parts: {
  components: string[];
  created: number;
  keyId: string;
  nonce: string;
}): string {
  const componentList = parts.components.map((component) => `"${component}"`).join(" ");
  return `(${componentList});created=${parts.created};keyid="${parts.keyId}";alg="ed25519";nonce="${parts.nonce}"`;
}

function resolveComponentValue(component: string, context: {
  method: string;
  url: URL;
  headers: Headers;
}): string {
  if (component === "@method") {
    // RFC 9421 derived component @method is lower-case.
    return context.method.toLowerCase();
  }
  if (component === "@target-uri") {
    return context.url.toString();
  }

  const headerValue = context.headers.get(component);
  if (!headerValue) {
    throw new Error(`Missing required signed header: ${component}`);
  }
  return headerValue;
}

function buildSigningBase(context: {
  components: string[];
  method: string;
  url: URL;
  headers: Headers;
  signatureParams: string;
}): string {
  const lines = context.components.map((component) => {
    const value = resolveComponentValue(component, {
      method: context.method,
      url: context.url,
      headers: context.headers,
    });
    return `"${component}": ${value}`;
  });

  lines.push(`"@signature-params": ${context.signatureParams}`);
  return lines.join("\n");
}

function parsePublicKey(publicKey: string): Uint8Array {
  if (!publicKey.startsWith("ed25519:")) {
    throw new Error("Unsupported public key format");
  }
  return new Uint8Array(Buffer.from(publicKey.slice("ed25519:".length), "base64"));
}

function hasComponent(components: string[], expected: string): boolean {
  return components.some((component) => component.trim() === expected);
}

function hasValidSignedComponentSet(
  components: string[],
  hasBody: boolean,
): boolean {
  const expected = hasBody
    ? REQUIRED_COMPONENTS_WITH_BODY
    : REQUIRED_COMPONENTS_NO_BODY;
  if (components.length !== expected.length) {
    return false;
  }
  for (let index = 0; index < expected.length; index += 1) {
    if (components[index] !== expected[index]) {
      return false;
    }
  }
  return true;
}

export function encodeCertificateHeader(certificate: SigilumCertificate): string {
  return Buffer.from(JSON.stringify(certificate), "utf8").toString("base64url");
}

export function decodeCertificateHeader(value: string): SigilumCertificate {
  const decoded = Buffer.from(value, "base64url").toString("utf8");
  return JSON.parse(decoded) as SigilumCertificate;
}

function parseSignatureInputHeader(value: string): SignatureInputParts {
  const match = value.match(
    /^sig1=\(([^)]*)\);created=(\d+);keyid="([^"]+)";alg="([^"]+)";nonce="([^"]+)"$/,
  );

  if (!match) {
    throw new Error("Invalid Signature-Input header format");
  }

  const rawComponents = match[1];
  const createdRaw = match[2];
  const keyId = match[3];
  const alg = match[4];
  const nonce = match[5];
  if (!rawComponents || !createdRaw || !keyId || !alg || !nonce) {
    throw new Error("Invalid Signature-Input header parts");
  }

  const components = rawComponents
    .trim()
    .split(/\s+/)
    .filter(Boolean)
    .map((token) => {
      if (!/^"[^"]+"$/.test(token)) {
        throw new Error(`Invalid component in Signature-Input: ${token}`);
      }
      return token.slice(1, -1);
    });

  return {
    components,
    created: Number(createdRaw),
    keyId,
    alg,
    nonce,
  };
}

function parseSignatureHeader(value: string): Uint8Array {
  const match = value.match(/^sig1=:([^:]+):$/);
  if (!match) {
    throw new Error("Invalid Signature header format");
  }
  const encoded = match[1];
  if (!encoded) {
    throw new Error("Invalid Signature header value");
  }
  return new Uint8Array(Buffer.from(encoded, "base64"));
}

function computeContentDigest(bodyBytes: Uint8Array): string {
  const digest = createHash("sha256").update(bodyBytes).digest("base64");
  return `sha-256=:${digest}:`;
}

export function signHttpRequest(
  identity: SigilumIdentity,
  request: SignRequestInput,
): SignedRequest {
  const url = normalizeUrl(request.url);
  const method = (request.method ?? "GET").toUpperCase();
  const headers = new Headers((request.headers ?? {}) as any);
  const bodyBytes = normalizeBody(request.body);

  if (bodyBytes && bodyBytes.length > 0) {
    headers.set("content-digest", computeContentDigest(bodyBytes));
  }

  headers.set("sigilum-namespace", identity.namespace);
  const resolvedSubject =
    request.subject?.trim() ||
    headers.get("sigilum-subject")?.trim() ||
    identity.namespace;
  headers.set("sigilum-subject", resolvedSubject);
  headers.set("sigilum-agent-key", identity.publicKey);
  headers.set("sigilum-agent-cert", encodeCertificateHeader(identity.certificate));

  const components = bodyBytes && bodyBytes.length > 0
    ? [
      "@method",
      "@target-uri",
      "content-digest",
      "sigilum-namespace",
      "sigilum-subject",
      "sigilum-agent-key",
      "sigilum-agent-cert",
    ]
    : ["@method", "@target-uri", "sigilum-namespace", "sigilum-subject", "sigilum-agent-key", "sigilum-agent-cert"];

  const created = request.created ?? Math.floor(Date.now() / 1000);
  const nonce = request.nonce ?? randomUUID();
  const signatureParams = buildSignatureParams({
    components,
    created,
    keyId: identity.keyId,
    nonce,
  });

  const signingBase = buildSigningBase({
    components,
    method,
    url,
    headers,
    signatureParams,
  });

  const signatureBytes = ed.sign(
    new TextEncoder().encode(signingBase),
    identity.privateKey,
  );

  headers.set("signature-input", `sig1=${signatureParams}`);
  headers.set("signature", `sig1=:${Buffer.from(signatureBytes).toString("base64")}:`);

  return {
    url: url.toString(),
    method,
    headers,
    body: request.body ?? null,
  };
}

export function verifyHttpSignature(
  request: VerifySignatureInput,
): VerifySignatureResult {
  const invalid = (code: string, reason: string): VerifySignatureResult => ({
    valid: false,
    code,
    reason,
  });

  try {
    const headers = new Headers(request.headers as any);
    const signatureInput = headers.get("signature-input");
    const signatureHeader = headers.get("signature");

    if (!signatureInput || !signatureHeader) {
      return invalid("SIG_MISSING_SIGNATURE_HEADERS", "Missing Signature-Input or Signature header");
    }

    let parsedInput: SignatureInputParts;
    try {
      parsedInput = parseSignatureInputHeader(signatureInput);
    } catch (error) {
      return invalid(
        "SIG_SIGNATURE_INPUT_INVALID",
        error instanceof Error ? error.message : String(error),
      );
    }
    if (!Number.isFinite(parsedInput.created) || parsedInput.created <= 0) {
      return invalid("SIG_CREATED_INVALID", "Invalid Signature-Input created timestamp");
    }
    if (parsedInput.alg.toLowerCase() !== "ed25519") {
      return invalid("SIG_ALGORITHM_UNSUPPORTED", "Unsupported signature algorithm");
    }

    const strict = request.strict;
    if (strict?.maxAgeSeconds != null) {
      const now = strict.now ?? Math.floor(Date.now() / 1000);
      const age = now - parsedInput.created;
      if (age < 0 || age > strict.maxAgeSeconds) {
        return invalid("SIG_TIMESTAMP_OUT_OF_RANGE", "Signature expired or not yet valid");
      }
    }
    if (strict?.nonceStore) {
      if (strict.nonceStore.has(parsedInput.nonce)) {
        return invalid("SIG_REPLAY_DETECTED", "Replay detected: nonce already seen");
      }
      strict.nonceStore.add(parsedInput.nonce);
    }

    let signature: Uint8Array;
    try {
      signature = parseSignatureHeader(signatureHeader);
    } catch (error) {
      return invalid(
        "SIG_SIGNATURE_HEADER_INVALID",
        error instanceof Error ? error.message : String(error),
      );
    }
    const method = request.method.toUpperCase();
    const url = normalizeUrl(request.url);
    const signatureParams = buildSignatureParams({
      components: parsedInput.components,
      created: parsedInput.created,
      keyId: parsedInput.keyId,
      nonce: parsedInput.nonce,
    });

    let signingBase: string;
    try {
      signingBase = buildSigningBase({
        components: parsedInput.components,
        method,
        url,
        headers,
        signatureParams,
      });
    } catch (error) {
      return invalid(
        "SIG_SIGNING_BASE_INVALID",
        error instanceof Error ? error.message : String(error),
      );
    }

    const certificateHeader = headers.get("sigilum-agent-cert");
    if (!certificateHeader) {
      return invalid("SIG_CERT_HEADER_MISSING", "Missing sigilum-agent-cert header");
    }
    let certificate: SigilumCertificate;
    try {
      certificate = decodeCertificateHeader(certificateHeader);
    } catch (error) {
      return invalid(
        "SIG_CERT_HEADER_INVALID",
        error instanceof Error ? error.message : String(error),
      );
    }

    if (!verifyCertificate(certificate)) {
      return invalid("SIG_CERT_INVALID", "Invalid agent certificate");
    }

    const publicKeyHeader = headers.get("sigilum-agent-key");
    if (!publicKeyHeader) {
      return invalid("SIG_KEY_HEADER_MISSING", "Missing sigilum-agent-key header");
    }

    if (certificate.publicKey !== publicKeyHeader) {
      return invalid("SIG_KEY_MISMATCH", "Certificate public key mismatch");
    }

    const namespaceHeader = headers.get("sigilum-namespace");
    if (!namespaceHeader || namespaceHeader !== certificate.namespace) {
      return invalid("SIG_NAMESPACE_MISMATCH", "Namespace header mismatch");
    }
    const subjectHeader = headers.get("sigilum-subject")?.trim();
    if (!subjectHeader) {
      return invalid("SIG_SUBJECT_MISSING", "Missing sigilum-subject header");
    }
    if (!hasComponent(parsedInput.components, "sigilum-subject")) {
      return invalid("SIG_SUBJECT_COMPONENT_MISSING", "Missing sigilum-subject in signed components");
    }

    const bodyBytes = normalizeBody(request.body);
    const hasBody = Boolean(bodyBytes && bodyBytes.length > 0);
    if (!hasValidSignedComponentSet(parsedInput.components, hasBody)) {
      return invalid("SIG_SIGNED_COMPONENTS_INVALID", "Invalid signed component set");
    }

    if (request.expectedNamespace && request.expectedNamespace !== namespaceHeader) {
      return invalid(
        "SIG_EXPECTED_NAMESPACE_MISMATCH",
        `Namespace mismatch: expected ${request.expectedNamespace}, got ${namespaceHeader}`,
      );
    }
    if (request.expectedSubject && request.expectedSubject !== subjectHeader) {
      return invalid(
        "SIG_EXPECTED_SUBJECT_MISMATCH",
        `Subject mismatch: expected ${request.expectedSubject}, got ${subjectHeader}`,
      );
    }

    if (parsedInput.keyId !== certificate.keyId) {
      return invalid("SIG_KEY_ID_MISMATCH", "keyid mismatch");
    }

    if (bodyBytes && bodyBytes.length > 0) {
      const expectedDigest = computeContentDigest(bodyBytes);
      if (headers.get("content-digest") !== expectedDigest) {
        return invalid("SIG_CONTENT_DIGEST_MISMATCH", "Content digest mismatch");
      }
    }

    let publicKey: Uint8Array;
    try {
      publicKey = parsePublicKey(publicKeyHeader);
    } catch (error) {
      return invalid(
        "SIG_KEY_HEADER_INVALID",
        error instanceof Error ? error.message : String(error),
      );
    }
    const valid = ed.verify(
      signature,
      new TextEncoder().encode(signingBase),
      publicKey,
    );

    if (!valid) {
      return invalid("SIG_VERIFICATION_FAILED", "Signature verification failed");
    }

    return {
      valid: true,
      namespace: certificate.namespace,
      subject: subjectHeader,
      keyId: certificate.keyId,
    };
  } catch (error) {
    return invalid("SIG_INTERNAL_ERROR", error instanceof Error ? error.message : String(error));
  }
}
